### PR TITLE
Bigtable: table deletion protection support

### PR DESCRIPTION
**Pull request by kimihrr**
- Add `deletion_protection` field to `resourceBigtableTable`
- `deletion_protection` should be string type as we need to determine if the field is given or not
- Deletion protection support while creating a table, reading a table and updating a table using terraform
- Adding supporting unit tests for different scenarios while creating a table with enabled deletion protection or disabled deletion protection.

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
bigtable: supported table deletion protection in terraform

```

### CODE REVIEW

File: mmv1/third_party/terraform/resources/resource_bigtable_table.go [hunk 1]
```diff
@@ -76,6 +77,15 @@ func resourceBigtableTable() *schema.Resource {
 				ForceNew:    true,
 				Description: `The ID of the project in which the resource belongs. If it is not provided, the provider project is used.`,
 			},
+
+			"deletion_protection": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
```
Comment (kevinsi4508) [inline mmv1/third_party/terraform/resources/resource_bigtable_table.go:84]: Add a comment here to say if not provided use API default.

File: mmv1/third_party/terraform/resources/resource_bigtable_table.go [hunk 2]
```diff
@@ -109,6 +119,15 @@ func resourceBigtableTableCreate(d *schema.ResourceData, meta interface{}) error
...
 	tblConf := bigtable.TableConf{TableID: tableId}
 
+	// DeletionProtection can be protected or unprotected
+	// Check if deletion protection is given
+	deletionProtection := d.Get("deletion_protection")
+	if deletionProtection == "PROTECTED" {
+		tblConf.DeletionProtection = bigtable.Protected
+	} else if deletionProtection == "UNPROTECTED" {
+		tblConf.DeletionProtection = bigtable.Unprotected
+	}
```
Comment (kevinsi4508) [inline mmv1/third_party/terraform/resources/resource_bigtable_table.go:129]: Add comment here: If not set, d.Get("deletion_protection") will return empty and the value will be set to the API default.

File: mmv1/third_party/terraform/resources/resource_bigtable_table.go [hunk 3]
```diff
@@ -188,6 +207,17 @@ func resourceBigtableTableRead(d *schema.ResourceData, meta interface{}) error {
...
+	deletionProtection := table.DeletionProtection
+	if deletionProtection == bigtable.Protected {
+		if err := d.Set("deletion_protection", "PROTECTED"); err != nil {
+			return fmt.Errorf("Error setting deletion_protection: %s", err)
+		}
+	} else if deletionProtection == bigtable.Unprotected {
+		if err := d.Set("deletion_protection", "UNPROTECTED"); err != nil {
+			return fmt.Errorf("Error setting deletion_protection: %s", err)
+		}
+	}
```
Comment (kevinsi4508) [inline mmv1/third_party/terraform/resources/resource_bigtable_table.go:220]: Can deletionProtection be something else? Maybe return an error if this happens?

File: mmv1/third_party/terraform/tests/resource_bigtable_table_test.go [hunk 4]
```diff
@@ -86,6 +87,144 @@ func TestAccBigtableTable_family(t *testing.T) {
 	})
 }
 
+func TestAccBigtableTable_deletionProtection_enabled(t *testing.T) {
```
Comment (kevinsi4508) [inline mmv1/third_party/terraform/tests/resource_bigtable_table_test.go:90]: nit: "_protected"?

File: mmv1/third_party/terraform/tests/resource_bigtable_table_test.go [hunk 5]
```diff
@@ -86,6 +87,144 @@ func TestAccBigtableTable_family(t *testing.T) {
...
+				ExpectError: regexp.MustCompile(".*deletion protection field is set to true.*"),
+			},
+			{
+				ResourceName:            "google_bigtable_instance.instance",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"deletion_protection", "instance_type"},
+			},
+			{
+				Config: testAccBigtableTable_family(instanceName, tableName, family),
```
Comment (kevinsi4508) [inline mmv1/third_party/terraform/tests/resource_bigtable_table_test.go:123]: What does this do? Maybe add a comment to explain?

**Review by kevinsi4508 — COMMENTED** [review]
Could add some comments in the test steps? Hard to follow. thanks!~

File: mmv1/third_party/terraform/tests/resource_bigtable_table_test.go [hunk 6]
```diff
@@ -86,6 +87,144 @@ func TestAccBigtableTable_family(t *testing.T) {
...
+			},
+			{
+				Config:      testAccBigtableTable(instanceName, tableName),
+				ExpectError: regexp.MustCompile(".*deletion protection field is set to true.*"),
+			},
+			{
+				ResourceName:            "google_bigtable_table.table",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"column_family"},
```
[the remark this answers is not in the document]
Comment (ScottSuarez) [reply mmv1/third_party/terraform/tests/resource_bigtable_table_test.go:142]: The expected state is based off the configuration. Since there was an error in the application of the configuration it makes sense not to verify the deployment. We can remove this step.

File: mmv1/third_party/terraform/tests/resource_bigtable_table_test.go [hunk 7]
```diff
@@ -86,6 +87,156 @@ func TestAccBigtableTable_family(t *testing.T) {
...
+			{
+				Config:      testAccBigtableTable_destroyTable(instanceName),
+				ExpectError: regexp.MustCompile(".*deletion protection field is set to true.*"),
+			},
+			{
+				ResourceName:            "google_bigtable_instance.instance",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"deletion_protection", "instance_type"},
+			},
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/tests/resource_bigtable_table_test.go:123]: this step can be removed.

**Review by ScottSuarez — CHANGES_REQUESTED** [review]
requesting changes clarifying end user scenarios, error handling, and comment changes.

File: mmv1/third_party/terraform/resources/resource_bigtable_table.go [hunk 8]
```diff
@@ -76,6 +77,16 @@ func resourceBigtableTable() *schema.Resource {
 				ForceNew:    true,
 				Description: `The ID of the project in which the resource belongs. If it is not provided, the provider project is used.`,
 			},
+
+			// If not provided, deletion protection will be set to the API default value
```
[the remark this answers is not in the document]
Comment (kevinsi4508) [reply mmv1/third_party/terraform/resources/resource_bigtable_table.go:81]: +1 to what Scott said. We can add something like this to the field description: If not provided, will currently set to `false` in the API.

File: mmv1/third_party/terraform/tests/resource_bigtable_table_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kevinsi4508) [reply mmv1/third_party/terraform/tests/resource_bigtable_table_test.go:142]: We can either remove it or keep it. It could be confusing if we keep it. We may have add a comment to explain why this is happening. It is an indication shows column_family change was not properly applied.

Either way, the goal is rather to verify the report infrastructure resource still have the column family.

File: mmv1/third_party/terraform/resources/resource_bigtable_table.go [hunk 9]
```diff
@@ -76,6 +77,16 @@ func resourceBigtableTable() *schema.Resource {
...
 			},
+
+			// If not provided, currently deletion protection will be set to false in the API
+			"deletion_protection": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validation.StringInSlice([]string{"PROTECTED", "UNPROTECTED"}, false),
+				Elem:         &schema.Schema{Type: schema.TypeString},
+				Description:  `A field to make the table protected against data loss i.e. when set to PROTECTED, deleting the table, the column families in the table, and the instance containing the table would be prohibited.`,
```
[the remark this answers is not in the document]
Comment (kimihrr) [reply mmv1/third_party/terraform/resources/resource_bigtable_table.go]: I added in the description. Thanks

File: mmv1/third_party/terraform/tests/resource_bigtable_table_test.go [hunk 10]
```diff
@@ -86,6 +87,125 @@ func TestAccBigtableTable_family(t *testing.T) {
...
+				ResourceName:      "google_bigtable_table.table",
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			// it is not possible to delete the table because of deletion protection equals to protected
+			{
+				Config:      testAccBigtableTable_destroyTable(instanceName),
+				ExpectError: regexp.MustCompile(".*deletion protection field is set to true.*"),
+			},
+			// changing the deletion protection field to unprotected
```
Comment (kevinsi4508) [inline mmv1/third_party/terraform/tests/resource_bigtable_table_test.go:196]: nit: changing the deletion protection field to unprotected so that the sources can properly be destroyed.

**Review by kevinsi4508 — APPROVED** [review]
LGTM

File: mmv1/third_party/terraform/resources/resource_bigtable_table.go [hunk 11]
```diff
@@ -76,6 +77,15 @@ func resourceBigtableTable() *schema.Resource {
...
 				Description: `The ID of the project in which the resource belongs. If it is not provided, the provider project is used.`,
 			},
+
+			"deletion_protection": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ValidateFunc: validation.StringInSlice([]string{"PROTECTED", "UNPROTECTED"}, false),
+				Elem:         &schema.Schema{Type: schema.TypeString},
+				Description:  `A field to make the table protected against data loss i.e. when set to PROTECTED, deleting the table, the column families in the table, and the instance containing the table would be prohibited. If not provided, currently deletion protection will be set to false as it is the API default value.`,
```
Comment (ScottSuarez) [inline mmv1/third_party/terraform/resources/resource_bigtable_table.go:87]:
```suggestion
				Description:  `A field to make the table protected against data loss i.e. when set to PROTECTED, deleting the table, the column families in the table, and the instance containing the table would be prohibited. If not provided, currently deletion protection will be set to `UNPROTECTED` as it is the API default value.`,
```

Comment (kimihrr) [reply mmv1/third_party/terraform/resources/resource_bigtable_table.go]: Done!

**Review by ScottSuarez — APPROVED** [review]
